static/app/utils/useRequest.tsx

```diff
@@ -0,0 +1,254 @@
+import {useState} from 'react';
+
+import {Client} from 'app/api';
+import OrganizationStore from 'app/stores/organizationStore';
+import {useLegacyStore} from 'app/stores/useLegacyStore';
+import parseLinkHeader from 'app/utils/parseLinkHeader';
+import RequestError from 'app/utils/requestError/requestError';
+import useApi from 'app/utils/useApi';
+
+type State = {
+  data: null | any;
+  /**
+   * This is state for when fetching data from API
+   */
+  loading: boolean;
+  /**
+   * The error that occurred if fetching failed
+   */
+  error: null | RequestError;
+  /**
+   * Indicates that Team results (from API) are paginated and there are more
+   * Teams that are not in the initial response.
+   */
+  hasMore: null | boolean;
+  /**
+   * The last query we searched. Used to validate the cursor
+   */
+  lastSearch: null | string;
+  /**
+   * Pagination
+   */
+  nextCursor?: null | string;
+};
+
+export type Result = {
+  /**
+   * This is an action provided to consumers for them to update the current
+   * result set using a simple search query.
+   *
+   * Will always add new options into the store.
+   */
+  onSearch: (searchTerm: string) => Promise<void>;
+  nextPage: () => Promise<void>;
+  get: () => Promise<void>;
+  post: (_) => Promise<void>;
+  update: (_) => Promise<void>;
+  del: () => Promise<void>;
+} & State;
+
+type Options = {
+  /**
+   * Number of repositories to return
+   */
+  limit?: number;
+};
+
+type FetchRequestOptions = {
+  limit?: Options['limit'];
+  cursor?: State['nextCursor'];
+  search?: State['lastSearch'];
+  lastSearch?: State['lastSearch'];
+};
+
+/**
+ * Helper function to actually load repositories
+ */
+async function fetch(
+  api: Client,
+  url: string,
+  {search, limit, lastSearch, cursor}: FetchRequestOptions = {}
+) {
+  const query: {
+    query?: string;
+    cursor?: typeof cursor;
+    per_page?: number;
+  } = {};
+
+  if (search) {
+    query.query = `${query.query ?? ''} ${search}`.trim();
+  }
+
+  const isSameSearch = lastSearch === search || (!lastSearch && !search);
+
+  if (isSameSearch && cursor) {
+    query.cursor = cursor;
+  }
+
+  if (limit !== undefined) {
+    query.per_page = limit;
+  }
+
+  let hasMore: null | boolean = false;
+  let nextCursor: null | string = null;
+  const [data, , resp] = await api.requestPromise(url, {
+    includeAllArgs: true,
+    query,
+  });
+
+  const pageLinks = resp?.getResponseHeader('Link');
+  if (pageLinks) {
+    const paginationObject = parseLinkHeader(pageLinks);
+    hasMore = paginationObject?.next?.results || paginationObject?.previous?.results;
+    nextCursor = paginationObject?.next?.cursor;
+  }
+
+  return {results: data, hasMore, nextCursor};
+}
+
+function useRequest(
```

Comment:
Can you please open this as it's own PR so we can review it away from useRepositories?